### PR TITLE
Add GroupWeakAddr.

### PR DESCRIPTION
**Pull request by alvardes**
This is useful for when you might have some group of "sibilings" of "child" actors that you want to notify, but don't want to extend their lifetime by holding an `Addr` of them.

Example could be when you have a `Listener`/`Server` which for a particular message/event creates a "child" actor.

The "child" actors could have a shorter lifespan than the `Listener` but they shouldn't outlive the `Listener`. In that case the `Listener` might want to store `WeakAddr`s for all of the "child" actors created and upon stopping it will send a message to all still existing "child" to notify them that they must shut down.

The reason for the pull request is that for this purpose you cannot use a simple `HashSet<WeakAddr<A>>` because `WeakAddr` doesn't (and cannot easily) implement `Hash` and `Eq`.

This PR's `GroupWeakAddr` internally uses the `PtrWeakHashSet` from the external crate `weak_table`.

All reviews are welcome. I am particularly uncertain about what is the best name for this struct so all suggestions are appreciated.

### CODE REVIEW

**Comment by alvardes** [comment]
Can I get a review @Jonathas-Conceicao?

**Comment by Jonathas-Conceicao** [comment]
@alvardes I'll try to find some time to carefully review this latter today.

File: src/address/group.rs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    fn view(&self) -> Option<Self::Strong> {
+        self.upgrade().map(|addr| WeakTableAddrWrap(addr))
+    }
+}
+
+impl<A: Actor> Clone for WeakTableWeakAddrWrap<A> {
+    fn clone(&self) -> WeakTableWeakAddrWrap<A> {
+        WeakTableWeakAddrWrap(self.0.clone())
+    }
+}
```
Comment (Jonathas-Conceicao) [inline src/address/group.rs:47]: Looks like this implementation can be derived. Less code to maintain is generally good.

File: src/address/group.rs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+pub struct GroupWeakAddr<A: Actor> {
+    set: PtrWeakHashSet<WeakTableWeakAddrWrap<A>>,
+}
+
+impl<A: Actor> Clone for GroupWeakAddr<A> {
+    fn clone(&self) -> GroupWeakAddr<A> {
+        GroupWeakAddr {
+            set: self.set.clone(),
+        }
+    }
```
Comment (Jonathas-Conceicao) [inline src/address/group.rs:62]: Use derive from this too.

File: src/address/group.rs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+}
+
+impl<A: Actor> PartialEq for GroupWeakAddr<A> {
+    fn eq(&self, other: &Self) -> bool {
+        self.set == other.set
+    }
+}
+
+impl<A: Actor> Eq for GroupWeakAddr<A> {}
```
Comment (Jonathas-Conceicao) [inline src/address/group.rs:92]: These can be derived too.

**Comment by Jonathas-Conceicao** [comment]
The general idea makes sense to me, but I'm not particularly good with names to help, comments from other contributors would be welcome.

Adding tests for this new tool is also important so we avoid any regression bugs, and they might work as a usage example as well. Also please update the CHANGES.md to mention the new struct.

**Comment by alvardes** [comment]
Actually while trying to write tests for this I realized that this doesn't seem to work. :(. I tried to make this work but I don't have a lot of free time to figure out whats wrong right now so I will close this for now.

**Comment by fakeshadow** [comment]
Please let me know if I get your idea. I believe something like this would be a suitable solution for your use case.
```rust
use std::any::{Any, TypeId};
use std::collections::HashMap;

use actix::prelude::*;
use actix_web::{get, App, HttpServer};

struct Host {
    children: HashMap<TypeId, Vec<Box<dyn Any>>>,
}

impl Host {
    fn add<M>(&mut self, recipient: Recipient<M>)
    where
        M: Message + Send + 'static,
        M::Result: Send,
    {
        let id = TypeId::of::<M>();

        self.children
            .entry(id)
            .or_insert(vec![])
            .push(Box::new(recipient));
    }
}

impl Actor for Host {
    type Context = Context<Self>;
}

#[derive(Default)]
struct Child1;

impl Actor for Child1 {
    type Context = Context<Self>;
}

#[derive(Default)]
struct Child2;

impl Actor for Child2 {
    type Context = Context<Self>;
}

#[derive(Clone)]
struct Broadcast;

#[derive(Clone)]
struct Register<M>(Recipient<M>)
where
    M: Message + Send + 'static,
    M::Result: Send;

impl Message for Broadcast {
    type Result = ();
}

impl<M> Message for Register<M>
where
    M: Message + Send + 'static,
    M::Result: Send,
{
    type Result = ();
}

impl<M> Handler<Register<M>> for Host
    where
        M: Message + Send + 'static,
        M::Result: Send,
{
    type Result = ();

    fn handle(&mut self, msg: Register<M>, _ctx: &mut Context<Self>) -> Self::Result {
        self.add(msg.0);
    }
}

impl Handler<Broadcast> for Host {
    type Result = ();

    fn handle(&mut self, msg: Broadcast, _ctx: &mut Context<Self>) -> Self::Result {
        self.children
            .get_mut(&TypeId::of::<Broadcast>())
            .unwrap()
            .into_iter()
            .for_each(|obj| {
                obj.downcast_mut::<Recipient<Broadcast>>()
                    .unwrap()
                    .do_send(msg.clone())
                    .unwrap()
            });
    }
}

impl Handler<Broadcast> for Child1 {
    type Result = ();

    fn handle(&mut self, _msg: Broadcast, _ctx: &mut Context<Self>) -> Self::Result {
        println!("Child1: we got message from master")
    }
}

impl Handler<Broadcast> for Child2 {
    type Result = ();

    fn handle(&mut self, _msg: Broadcast, _ctx: &mut Context<Self>) -> Self::Result {
        println!("Child2: we got message from master")
    }
}

#[actix_rt::main]
async fn main() -> std::io::Result<()> {
    let master = Host::create(|_| {
        Host {
            children: Default::default()
        }
    });

    let child1 = Child1::start_default();

    let child2 = Child2::start_default();

    master.send(Register(child1.recipient())).await.unwrap();

    master.send(Register(child2.recipient())).await.unwrap();

    master.send(Broadcast).await.unwrap();

    actix_rt::time::delay_for(std::time::Duration::from_secs(1)).await;

    Ok(())
}
```
By using Recipent instead of Addr you can have multiple children actors implement different kinds of messages. 
Yes there is some runtime cost for an extra layer of trait object and there is no WeakRecipient but I believe they are not hard to implement.

**Pull request closed without merging**